src/compas/datastructures/volmesh/core/halfface.py

```diff
@@ -476,7 +479,7 @@ def add_halfface(self, vertices, fkey=None, attr_dict=None, **kwattr):
                 self._plane[w][v][u] = None
         return fkey
 
-    def add_cell(self, faces, ckey=None, attr_dict=None, **kwattr):
+    def add_cell(self, faces, ckey=None, attr_dict=None, fkeys=None, **kwattr):
```

Comment:
Hey @tomvanmele Could you take a look at this change here: `add_cell` used to append new halffaces to the class, however in the case of `data setter` , the halffaces are already added initially, calling `add_cell` results in having duplicated halffaces.  So there needs to be an optional argument to supply an existing fkey instead of generate a new one.

Reply:
in my opinion things can be much simpler
```python
    @property
    def data(self):
        """dict: A data dict representing the volmesh data structure for serialisation.
        """
        # edge_data = {}
        # for e in self._edge_data:
        #     edge_data[repr(e)] = self._edge_data[e]
        data = {
            'attributes': self.attributes,
            'dva': self.default_vertex_attributes,
            'dea': self.default_edge_attributes,
            'dfa': self.default_face_attributes,
            'dca': self.default_cell_attributes,
            'vertex': self._vertex,
            'halfface': self._halfface,
            'cell': self._cell,
            'plane': self._plane,
            'edge_data': self.edge_data,
            'face_data': self._face_data,
            'cell_data': self._cell_data,
            'max_vertex': self._max_vertex,
            'max_face': self._max_face,
            'max_cell': self._max_cell}
        return data

    @data.setter
    def data(self, data):
        attributes = data.get('attributes') or {}
        dva = data.get('dva') or {}
        dea = data.get('dea') or {}
        dfa = data.get('dfa') or {}
        dca = data.get('dca') or {}
        vertex = data.get('vertex') or {}
        halfface = data.get('halfface') or {}
        cell = data.get('cell') or {}
        # plane = data.get('plane') or {}
        edge_data = data.get('edge_data') or {}
        face_data = data.get('face_data') or {}
        cell_data = data.get('cell_data') or {}
        max_vertex = data.get('max_vertex', -1)
        max_face = data.get('max_face', -1)
        max_cell = data.get('max_cell', -1)

        if not vertex or not halfface or not cell:
            return

        self.attributes.update(attributes)
        self.default_vertex_attributes.update(dva)
        self.default_edge_attributes.update(dea)
        self.default_face_attributes.update(dfa)
        self.default_cell_attributes.update(dca)

        self._vertex = {}
        self._halfface = {}
        self._cell = {}
        self._plane = {}
        self._edge_data = {}
        self._face_data = {}
        self._cell_data = {}

        for v in vertex:
            attr = vertex[v] or {}
            self.add_vertex(int(v), attr_dict=attr)

        # for f in halfface:
        #     attr = face_data.get(f) or {}
        #     self.add_halfface(halfface[f], fkey=int(f), attr_dict=attr)

        for c in cell:
            attr = cell_data.get(c) or {}
            faces = []
            for u in cell[c]:
                for v in cell[c][u]:
                    f = cell[c][u][v]
                    faces.append(halfface[f])
            self.add_cell(faces, ckey=int(c), attr_dict=attr)

        for e in edge_data:
            self._edge_data[e] = edge_data[e] or {}

        for f in face_data:
            self._face_data[f] = face_data[f] or {}

        self._max_vertex = max_vertex
        self._max_face = max_face
        self._max_cell = max_cell
```

Reply:
in the data getter edge data conversion can be commented out because in the new implementation all `edge_data` dict keys are strings by default.

this means the `literal_eval` on the setter side can also be omitted, but perhaps both can be left if for backward compatibility...

Reply:
the half-faces should indeed be created implicitly by the `add_cell` method, and the face identifiers from `f = cell[c][u][v]` have to be converted to strings before looking up the vertices in the half-face dict because of the implicit string conversion by JSON => `faces.append(halfface[str(f)])`.

other than that i don't see the need for additional conversions. all data not stored as dict keys will just retain its original type (int, mostly)

Reply:
the edge data and face data loops at the end could be avoided entirely if not for the potential empty dicts having been converted to lists by JSON

Reply:
the problem just lies in the difference between the data dict and the JSON serialized data dict. in the former halfface keys are still `int` and in the latter they will have been converted to `str`. however, we can simply solve this by storing the halffaces directly in a cell dict before serializing rather than in a cell dict that points at a halfface dict...

Reply:
```python
import compas
from compas.datastructures import VolMesh

vmesh1 = VolMesh.from_obj(compas.get('boxes.obj'))

data1 = vmesh1.to_data()

vmesh1.to_json('vmesh1.json')
vmesh1.from_json('vmesh1.json')

vmesh1.validate_data()
vmesh1.validate_json()

data1_ = vmesh1.to_data()

print(data1 == data1_)

vmesh2 = VolMesh.from_data(data1_)

vmesh2.validate_data()
vmesh2.validate_json()

data2 = vmesh2.to_data()

vmesh2.to_json('vmesh2.json')
vmesh2.from_json('vmesh2.json')

data2_ = vmesh2.to_data()

print(data2 == data2_)

print(data1 == data2)
```

Reply:
you also have to update the schemas for this to work, though...